CORE-8902 - Addon detection & start-up banner

driessamyn opened the pull request:
Prep work for being able to include Corda ENT modules.
This change includes:

- Printing a Corda start-up banner in all workers.
- OS implementation of this banner is included. This can be overridden by having a different implementation of the banner when one is loaded with a higher `ServiceRanking` (see OSGi test for example). This is how Corda ENT workers can display a different banner at start-up, without changing the worker code.
- Details of add-ons discovered at start-up will be printed to the console, after the start-up banner. This is how ENT add-ons can be loaded. Discovery works similar to the JDBC driver discovery.

Loading add-ons is tested in the OSGi test, but the discovery from bundles on disk will need to be tested when Corda ENT add-ons are available.

Example (OS) banner:
```
   ______               __
  / ____/     _________/ /___ _
 / /     __  / ___/ __  / __ `/
/ /___  /_/ / /  / /_/ / /_/ /
\____/     /_/   \__,_/\__,_/

--- Combined Worker (5.0.0.0-SNAPSHOT) ---
```

Example ENT banner (see PR #4 in the enterprise repo):

```
   ______               __         _____ _   _ _____ _____ ____  ____  ____  ___ ____  _____
  / ____/     _________/ /___ _   | ____| \ | |_   _| ____|  _ \|  _ \|  _ \|_ _/ ___|| ____|
 / /     __  / ___/ __  / __ `/   |  _| |  \| | | | |  _| | |_) | |_) | |_) || |\___ \|  _|
/ /___  /_/ / /  / /_/ / /_/ /    | |___| |\  | | | | |___|  _ <|  __/|  _ < | | ___) | |___
\____/     /_/   \__,_/\__,_/     |_____|_| \_| |_| |_____|_| \_\_|   |_| \_\___|____/|_____|

Combined Worker (5.0.0.0-SNAPSHOT)

*************************************************************************************************************************************
*  All rights reserved.                                                                                                             *
*  This software is proprietary to and embodies the confidential technology of R3 LLC ("R3").                                       *
*  Possession, use, duplication or dissemination of the software is authorized only pursuant to a valid written license from R3.    *
*  IF YOU DO NOT HAVE A VALID WRITTEN LICENSE WITH R3, DO NOT USE THIS SOFTWARE.                                                    *
*************************************************************************************************************************************


******************************************************** Available add-ons:  ********************************************************
*************************************************************************************************************************************
***************************************** HashiCorp Vault Configuration (5.0.0.0-SNAPSHOT)  *****************************************
    Provides HashiCorp vault integration with the Corda configuration system.

    R3 LLC ("R3")

    All rights reserved.
    This software is proprietary to and embodies the confidential technology of R3 LLC ("R3").
    Possession, use, duplication or dissemination of the software is authorized only pursuant to a valid written license from R3.
    IF YOU DO NOT HAVE A VALID WRITTEN LICENSE WITH R3, DO NOT USE THIS SOFTWARE.
*************************************************************************************************************************************
```

Commit:
Addon detection & start-up banner.

vkolomeyko reviewed (dismissed):
LGTM from @corda/http-rpc point of view.

The only thing I would add into description is a sample of how banner would look like, along with set of addons.

ronanbrowne reviewed (dismissed):
lgtm (build)

mnesbit commented on libs/application/banner/src/main/kotlin/net/corda/application/banner/CordaOsBanner.kt:
```diff
@@ -0,0 +1,79 @@
+@Component(service = [StartupBanner::class])
+class CordaOsBanner : StartupBanner {
+    override fun get(name: String, version: String): String {
+        val messages = listOf(
```
Please remove the 'jokes'. They are troublesome at best.

driessamyn replied:
It's a copy/paste from Corda 4 (except from the Emoji stuff as I didn't want to port that). Just ensuring feature parity ;)

But I'll remove them.

Commit:
Stop Corda being funny.

mnesbit commented on libs/application/addon/src/main/kotlin/net/corda/application/addon/CordaAddon.kt:
```diff
@@ -0,0 +1,17 @@
+
+interface CordaAddon {
+    val name: String
+    val licence: String
```
Are the licenses of all add-ons clearly printed at startup? They are the critical piece and need to be easily updateable as wording changes often. Historically, we have also have some releases that required a user acceptance before program continued and also discussions about having installable license code files. This might mean this should be `fun getLicences(codes: List<String>)` rather than a simple property, but maybe needs a short discussion

driessamyn replied:
There is an action with Product to define the requirements around licencing (e.g. do we licence add-ons individually or have a single "enterprise" licence). For now, we decided to keep it simple as we will likely just have 1 add-on for 5.0.

For now, it just does:

```
addons.forEach {
            logger.info("Addon available: ${it.name} (${it.version})")
            consolePrinter.printEmptyLine()
            consolePrinter.printPaddedLine("${it.name} (${it.version})")
            consolePrinter.printLeftPad("${it.description}")
            println()
            consolePrinter.printLeftPad("${it.vendor}")
            consolePrinter.printLeftPad("${it.licence}")
        }
```

So it should be easy enough to change. I mostly wanted to get to a point we can test out add-ons being loaded from an external folder.

mnesbit replied:
Yes I would like to be included in that product meeting

rick-r3 commented on osgi-framework-bootstrap/src/main/java/net/corda/osgi/framework/OSGiFrameworkMain.java:
```diff
@@ -206,4 +207,27 @@ private static Path getDbDriverDirectory(String[] args) {
+     * we can install them here.
+     */
+    private static Path getAddonDirectory(String[] args) {
+        final List<String> jdbcValues = Arrays.stream(args)
```
Mention of JDBC unexpected

driessamyn replied:
Thanks.
I refactored it a bit to reduce the duplication.

rick-r3 commented on applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/ApplicationBanner.kt:
```diff
@@ -0,0 +1,50 @@
+            consolePrinter.printEmptyLine()
+            consolePrinter.printPaddedLine("${it.name} (${it.version})")
+            consolePrinter.printLeftPad("${it.description}")
+            println()
```
Add and use `consolePrinter.println()` here ?  If `consolePrinter` does something else, you get a rogue empty line printing to the console.

rick-r3 commented on applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/ApplicationBanner.kt:
```diff
@@ -0,0 +1,50 @@
+        consolePrinter.printPaddedLine("Available add-ons:")
+
+        addons.forEach {
+            logger.info("Addon available: ${it.name} (${it.version})")
```
I think you want mostly the same content going to the logs.  Currently things like the description, vendor etc do not.  Should there be more `logger` lines here or should `consolePrinter` also echo to logs etc?

driessamyn replied:
I'm not sure what's best. 
It's common to log to `stoud` in a docker/Kubernetes environment (and that is our default), in most cases everything will go to log anyway, but the bits written to console directly may be thrown away by any log collectors are they will not conform to the logging format. 
From a logging perspective, I didn't think things like the licence and description would be useful, but it may be useful to know the version of the add-ons.

@davidcurrie , any thoughts?

rick-r3 replied:
Fair enough.  I don't feel strongly about it and obviously we can revise if we change our minds.  There's always a danger of further divergence and then something useful ends up missing from one or the other, but that probably isn't the case currently.

mnesbit replied:
I feel that banners aren't important in logs (assuming we still have out version info), but we will want the plugin types and versions listed to logs and we would want the license strings in the log file so nobody can say they didn't see them

rick-r3 commented on osgi-framework-bootstrap/src/main/java/net/corda/osgi/framework/OSGiFrameworkMain.java:
```diff
@@ -206,4 +207,27 @@ private static Path getDbDriverDirectory(String[] args) {
+            .collect(toUnmodifiableList());
+
+        if (!jdbcValues.isEmpty()) {
+            final String jdbcValue = jdbcValues.get(0);
```
More JDBC mentions.

Commit:
PR feedback.

Commit:
- Tweak console output
- Fix gradle project description
- Add add-on common start-up param so PicoCLI doesn't reject it.

mnesbit reviewed (approved):
Will do for now, but needs a bit of discussion with product before GA

vkolomeyko reviewed (approved):
LGTM

ronanbrowne reviewed (approved):
lgtm (build)
